addons/sale_coupon/models/sale_order.py

```diff
@@ -439,6 +439,21 @@ def _get_applied_programs(self):
         """
         return self.code_promo_program_id + self.no_code_promo_program_ids + self.applied_coupon_ids.mapped('program_id')
 
+    def _get_invoice_status(self):
+        # Handling of a specific situation: an order contains
+        # a product invoiced on delivery and a promo line invoiced
+        # on order. We would avoid having the invoice status 'to_invoice'
+        # if the created invoice will only contain the promotion line
+        super()._get_invoice_status()
+        for order in self.filtered(lambda order: order.invoice_status == 'to invoice'):
+            paid_lines = order._get_paid_order_lines()
+            if not any(line.invoice_status == 'to invoice' for line in paid_lines):
+                order.invoice_status = 'no'
+
+    def _check_invoice_vals(self, vals):
+        # Override to check if invoice lines are only promotion lines
+        return not all([val[2].get('coupon_program_id') for val in vals])
```

Comment:
`coupon_program_id` is only given when `sale_coupon_taxcloud` is installed, you cannot use this logic in the main `sale_coupon` module.